File: solace/cmd/play/Remove.java
package solace.cmd.play;

import solace.net.*;
import solace.util.*;
import solace.game.*;

/**
 * Remove command, allows players to remove equipment from their characters.
 * @author Ryan Sandor Richards
 */
public class Remove extends PlayCommand {
  public Remove(solace.game.Character ch) {
    super("remove", ch);
  }

  public boolean run(Connection c, String []params) {
    if (params.length < 2) {
      character.sendln("What would you like to remove?");
      return false;
    }

    if (character.getPlayState() == PlayState.SLEEPING) {
      character.sendln("You cannot remove equipment whilst alseep.");
      return false;
    }

    String itemName = params[1];
    Item removeItem = null;

    for (String slot : solace.game.Character.EQ_SLOTS) {
      Item item = character.getEquipment(slot);
      if (item == null) {
        continue;
      }
      if (item.hasName(itemName)) {
        removeItem = item;
        break;
      }
    }

    if (removeItem == null) {
      character.sendln("You are not wearing " + itemName + ".");
      return false;
    }

    String itemDesc = removeItem.get("description.inventory");

    try {
      character.unequip(removeItem);
      character.sendln("You removed " + itemDesc);
    }
    catch (NotEquipmentException ne) {
      // Shouldn't happen because of checks above...
      Log.error("Attempted to remove non-equipment item: " + itemDesc);
      ne.printStackTrace();
      character.sendln("You cannot remove " + itemDesc);
      return false;
    }
    catch (NoSuchItemException nse) {
      // Shouldn't happen because of checks above...
      Log.error("Attempted to remove unheld item: " + itemDesc);
      nse.printStackTrace();
      character.sendln("You cannot remove: " + itemDesc);
      return false;
    }

    return true;
  }
}

File: game/commands/inventory/remove.js
'use strict';

var EQ_SLOTS = Packages.solace.game.Character.EQ_SLOTS

/**
 * Remove command, allows players to remove equipment from their characters.
 * @author Ryan Sandor Richards
 */
Commands.add('remove', function (player, params) {
  if (player.isMobile()) {
    return player.sendln("Mobiles do not have equipment.");
  }
  var character = player.getCharacter();

  if (params.length < 2) {
    return character.sendln("What would you like to remove?");
  }

  if (character.isSleeping()) {
    return character.sendln("You cannot remove equipment whilst alseep.");
  }

  var itemName = params[1];
  var removeItem = null;
  EQ_SLOTS.forEach(function (slot) {
    var item = character.getEquipment(slot);
    if (item == null) return;
    if (item.hasName(itemName)) {
      removeItem = item;
    }
  });

  if (removeItem == null) {
    return character.sendln("You are not wearing " + itemName + ".");
  }

  var itemDesc = removeItem.get("description.inventory");
  try {
    character.unequip(removeItem);
    character.sendln("You removed " + itemDesc);
  } catch (e) {
    Log.error("Error when removing equipment " + itemDesc);
    character.sendln("You could not remove " + itemDesc);
  }
});
